File: frontend/src/types/events.ts
export type EventType =
  | 'pipeline_status'
  | 'task_created'
  | 'task_assigned'
  | 'task_updated'
  | 'agent_thinking'
  | 'agent_output'
  | 'code_generated'
  | 'review_result'
  | 'test_result'
  | 'error'
  | 'log'

interface BaseEvent {
  run_id: string
  timestamp: string
}

export interface PipelineStatusEvent extends BaseEvent {
  type: 'pipeline_status'
  data: { status: string }
}

export interface TaskCreatedEvent extends BaseEvent {
  type: 'task_created'
  data: { task_id: string; title: string; specialty_tags: string[] }
}

export interface TaskAssignedEvent extends BaseEvent {
  type: 'task_assigned'
  data: { task_id: string; developer: string; developer_emoji: string }
}

export interface TaskUpdatedEvent extends BaseEvent {
  type: 'task_updated'
  data: { task_id: string; status: string }
}

export interface AgentThinkingEvent extends BaseEvent {
  type: 'agent_thinking'
  data: { agent: string; message: string }
}

export interface AgentOutputEvent extends BaseEvent {
  type: 'agent_output'
  data: { agent: string; output: string }
}

export interface CodeGeneratedEvent extends BaseEvent {
  type: 'code_generated'
  data: { agent: string; files: string[] }
}

export interface ReviewResultEvent extends BaseEvent {
  type: 'review_result'
  data: { approved: boolean; comments_count: number }
}

export interface TestResultEvent extends BaseEvent {
  type: 'test_result'
  data: { passed: boolean; test_count: number }
}

export interface ErrorEvent extends BaseEvent {
  type: 'error'
  data: { message: string }
}

export interface LogEvent extends BaseEvent {
  type: 'log'
  data: { message: string }
}

export type PipelineEvent =
  | PipelineStatusEvent
  | TaskCreatedEvent
  | TaskAssignedEvent
  | TaskUpdatedEvent
  | AgentThinkingEvent
  | AgentOutputEvent
  | CodeGeneratedEvent
  | ReviewResultEvent
  | TestResultEvent
  | ErrorEvent
  | LogEvent

File: backend/app/ws/events.py
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any


class EventType(str, Enum):
    PIPELINE_STATUS = "pipeline_status"
    TASK_CREATED = "task_created"
    TASK_ASSIGNED = "task_assigned"
    TASK_UPDATED = "task_updated"
    AGENT_THINKING = "agent_thinking"
    AGENT_OUTPUT = "agent_output"
    CODE_GENERATED = "code_generated"
    REVIEW_RESULT = "review_result"
    TEST_RESULT = "test_result"
    ERROR = "error"
    LOG = "log"


@dataclass
class PipelineEvent:
    type: EventType
    run_id: str
    data: dict[str, Any] = field(default_factory=dict)
    timestamp: str = ""

    def __post_init__(self) -> None:
        if not self.timestamp:
            self.timestamp = datetime.now(timezone.utc).isoformat()

    def to_dict(self) -> dict:
        d = asdict(self)
        d["type"] = self.type.value
        return d
